chain-extensions/pallet-assets/src/lib.rs

```diff
@@ -0,0 +1,371 @@
+use pallet_assets::WeightInfo;
+use pallet_contracts::chain_extension::{
+    ChainExtension, Environment, Ext, InitState, RetVal, SysConfig,
+};
+use sp_runtime::traits::StaticLookup;
+use sp_runtime::DispatchError;
+use sp_std::marker::PhantomData;
+use sp_std::vec::Vec;
+
+enum AssetsFunc {
+    Create,
+    Transfer,
+    Mint,
+    Burn,
+    BalanceOf,
+    TotalSupply,
+    Allowance,
+    ApproveTransfer,
+    CancelApproval,
+    TransferApproved,
+    SetMetadata,
+    MetadataName,
+    MetadataSymbol,
+    MetadataDecimals,
+}
+
+impl TryFrom<u16> for AssetsFunc {
+    type Error = DispatchError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        match value {
+            1 => Ok(AssetsFunc::Create),
+            2 => Ok(AssetsFunc::Transfer),
+            3 => Ok(AssetsFunc::Mint),
+            4 => Ok(AssetsFunc::Burn),
+            5 => Ok(AssetsFunc::BalanceOf),
+            6 => Ok(AssetsFunc::TotalSupply),
+            7 => Ok(AssetsFunc::Allowance),
+            8 => Ok(AssetsFunc::ApproveTransfer),
+            9 => Ok(AssetsFunc::CancelApproval),
+            10 => Ok(AssetsFunc::TransferApproved),
+            11 => Ok(AssetsFunc::SetMetadata),
+            12 => Ok(AssetsFunc::MetadataName),
+            13 => Ok(AssetsFunc::MetadataSymbol),
+            14 => Ok(AssetsFunc::MetadataDecimals),
+            _ => Err(DispatchError::Other(
+                "PalletAssetsExtension: Unimplemented func_id",
+            )),
+        }
+    }
+}
+
+/// Pallet Assets chain extension.
+pub struct AssetsExtension<T, W>(PhantomData<(T, W)>);
+
+impl<T, W> Default for AssetsExtension<T, W> {
+    fn default() -> Self {
+        AssetsExtension(PhantomData)
+    }
+}
+
+impl<T, W> ChainExtension<T> for AssetsExtension<T, W>
+where
+    T: pallet_assets::Config + pallet_contracts::Config,
+    <<T as SysConfig>::Lookup as StaticLookup>::Source: From<<T as SysConfig>::AccountId>,
+    <T as SysConfig>::AccountId: From<[u8; 32]>,
+    W: weights::WeightInfo,
+{
+    fn call<E: Ext>(&mut self, env: Environment<E, InitState>) -> Result<RetVal, DispatchError>
+    where
+        E: Ext<T = T>,
+    {
+        let func_id = env.func_id().try_into()?;
+        let mut env = env.buf_in_buf_out();
+
+        match func_id {
+            AssetsFunc::Create => {
+                let (origin, id, admin, min_balance): (
+                    Origin,
+                    <T as pallet_assets::Config>::AssetId,
+                    T::AccountId,
+                    T::Balance,
+                ) = env.read_as()?;
+
+                let base_weight = <T as pallet_assets::Config>::WeightInfo::create();
+                env.charge_weight(base_weight)?;
+
+                let call_result = pallet_assets::Pallet::<T>::create(
+                    Origin::get_origin(&origin, env).into(),
+                    id.into(),
+                    admin.into(),
+                    min_balance,
+                );
```

Comment:
I believe this is missing the storage deposit limit check.

Creating a new asset will result in some native balance being reserved.
However, this won't be accounted for in the storage meter when compared to `storage_deposit_limit` :frowning_face: 

Looking at the chain-extension code, I feel like this is a bug in `pallet-contracts` since `storage_meter` isn't exposed. We also don't know how much balance was reserved as the result of successful `create` call.

Not sure how to tackle this.

Reply:
The create function is a payable method where you pass the amount needed. You can get this [amount with env](https://paritytech.github.io/ink/ink_env/fn.transferred_value.html).
[ink! message](https://github.com/swanky-dapps/chain-extension-contracts/blob/eb9b4f11f33c8b98156bbe4565a344f095b426e4/examples/assets/lib.rs#L24)
[test](https://github.com/swanky-dapps/chain-extension-contracts/blob/eb9b4f11f33c8b98156bbe4565a344f095b426e4/tests/assets.spec.ts#L51)

Reply:
Ok, but there is no guarantee that the amount provided will be sufficient to pay for the asset creation, right?

E.g. a caller could pass 10^-18 ASTR and make use of 1000 ASTR available in contract to create an asset.

Chain extension should calculate the price required (just an idea), and `pallet-assets` should expose some API via which this calculation can done (doesn't seem right to copy/paste logic from pallet into chain-extension).

Reply:
I can check [value_transferred](https://github.com/paritytech/substrate/blob/38c3e5ee10be760115a77a115ef0ac6ba381542b/frame/contracts/src/exec.rs#L243) and query the deposit to check if it's "equal" ?

Reply:
Sure, but with the changes to how the chain-extension will work (discussed yesterday), not allowing reuse of `caller` as origin to `pallet-assets` calls, then this problem is no more?

Reply:
It is still because the transferred balance come from caller but is balance of contract and only contract can pay for those fees

Reply:
Coding this makes me realize there is no need for this. As only caller will be origin of the extrinsic the contract will reserve the deposit.
If contract does not have enough balance reserve will fail and Err will be returned by the call.
If contracts does have the deposit it will be reserved.

Reply:
Yes, that makes sense. Don't see a use case when a caller would create an asset via smart contract.